source/Cosmos.HAL2/BlockDevice/ATAPI.cs

```diff
@@ -0,0 +1,187 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Cosmos.Core;
+
+namespace Cosmos.HAL.BlockDevice
+{
+    public class ATAPI : BlockDevice
+    {
+        /*
+         *  Capacity is worked out using the following:
+         *  Capacity = (Last LBA + 1) * Block Size (virtually always 2KB)
+         *
+         */
+        protected static Core.IOGroup.ATA IO;
+        private const UInt16 SectorSize = 2048;
+        private const UInt32 LBA = 0;
+        public static UInt16[] DataBuffer = new UInt16[256];
+        public static bool IRQReceived = false;
+        private byte[] ReadToc = { 0x43, 0, 1, 0, 0, 0, 0, 0, 12, 0x40, 0, 0 };
+
+        public static void Init()
+        {
+            Core.INTs.SetIrqHandler(0x0F, HandleIRQ);
+            IO.DeviceSelect.Byte = 0xA0;
+            IO.Control.Byte = 0x00;
+        }
+
+        public static void Test()
+        {
+            ReadBlock(1);
+            PrintBuffer();
+        }
+
+        public static void HandleIRQ(ref Cosmos.Core.INTs.IRQContext aContext)
+        {
+            Console.WriteLine("IRQ15");
+            IRQReceived = true;
+        }
+
+        public static void PrintBuffer()
+        {
+            for (int p = 0; p < 256; p++)
+            {
+                if (p != 255)
+                {
+                    Console.Write(DataBuffer[p].ToString() + ",");
+                }
+                else
+                {
+                    Console.WriteLine(DataBuffer[p].ToString());
+                }
+            }
+        }
+
+        public static void ReadBlock(uint lba)
```

Comment:
Maybe it is because is all debug / experimental code but the argument 'lba' is not used by the function?

Reply:
It's just experimental code for now, found it in an old driver when I was comparing which one worked and which didn't. It won't be included in the final part.

I was also a bit confused how to specify which block to read from the ATAPI drive (and if a ATAPI drive is a block device?)

https://github.com/CosmosOS/Cosmos/blob/fa4c223452a751f889ff7c53d55c7d8318b9acdb/source/Cosmos.HAL2/BlockDevice/BlockDevice.cs#L40